util/library_loader.h

```diff
@@ -0,0 +1,162 @@
+  virtual ~LibraryLoader() = default;
+
+  bool IsValid() const { return is_valid_; }
+
+  virtual void *GetEntryPoint(const char *function_name) = 0;
+
+ protected:
+  bool is_valid_;
+};
+
+class UnixLibraryLoader : public LibraryLoader {
+ public:
+  UnixLibraryLoader() = delete;
+
+  UnixLibraryLoader(const char *library_name);
+
+  virtual ~UnixLibraryLoader();
+
+  virtual void *GetEntryPoint(const char *function_name) override;
+
+  virtual size_t GetEntryPoints(std::map<std::string, void *> &functions);
+
+ protected:
+  void *dl_handle_;
+  std::string last_error_msg_;
+};
+
+class UnixLibCrypto : public UnixLibraryLoader {
+ public:
+  UnixLibCrypto();
+  virtual ~UnixLibCrypto() = default;
+
+  // _new & _free are ssl 1.1, replacing 1.0 _create & _destroy
+  using EVP_MD_CTX_new_t = EVP_MD_CTX *(*)(void);
+  using EVP_DigestInit_ex_t = int (*)(EVP_MD_CTX *ctx, const EVP_MD *type,
+                                      ENGINE *impl);
+  using EVP_sha1_t = const EVP_MD *(*)(void);
+  using EVP_DigestUpdate_t = int (*)(EVP_MD_CTX *ctx, const void *d,
+                                     size_t cnt);
+  using EVP_DigestFinal_ex_t = int (*)(EVP_MD_CTX *ctx, unsigned char *md,
+                                       unsigned int *s);
+  using EVP_MD_CTX_free_t = void (*)(EVP_MD_CTX *ctx);
+
+  EVP_MD_CTX *EVP_MD_CTX_new() const { return ctx_new_(); };
+
+  int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *impl) {
+    return digest_init_(ctx, type, impl);
+  }
+
+  const EVP_MD *EVP_sha1() { return sha1_(); }
+
+  int EVP_DigestUpdate(EVP_MD_CTX *ctx, const void *d, size_t cnt) {
+    return digest_update_(ctx, d, cnt);
+  }
+
+  int EVP_DigestFinal_ex(EVP_MD_CTX *ctx, unsigned char *md, unsigned int *s) {
+    return digest_final_(ctx, md, s);
+  }
+
+  void EVP_MD_CTX_free(EVP_MD_CTX *ctx) { ctx_free_(ctx); }
+
+  EVP_MD_CTX_free_t EVP_MD_CTX_free_ptr() { return ctx_free_; }
+
+  using RAND_bytes_t = int (*)(unsigned char *buf, int num);
+  using RAND_poll_t = int (*)();
+
+  int RAND_bytes(unsigned char *buf, int num) { return rand_bytes_(buf, num); }
+
+  int RAND_poll() { return rand_poll_(); }
+
+  using EVP_CIPHER_CTX_new_t = EVP_CIPHER_CTX *(*)(void);
+  using EVP_CIPHER_CTX_free_t = void (*)(EVP_CIPHER_CTX *ctx);
+  using EVP_EncryptInit_ex_t = int (*)(EVP_CIPHER_CTX *ctx,
+                                       const EVP_CIPHER *type, ENGINE *impl,
+                                       const unsigned char *key,
+                                       const unsigned char *iv);
+  using EVP_aes_256_ctr_t = const EVP_CIPHER *(*)(void);
+  using EVP_EncryptUpdate_t = int (*)(EVP_CIPHER_CTX *ctx, unsigned char *out,
+                                      int *outl, const unsigned char *in,
+                                      int inl);
+
+  EVP_CIPHER_CTX *EVP_CIPHER_CTX_new(void) const { return cipher_new_(); };
+
+  void EVP_CIPHER_CTX_free(EVP_CIPHER_CTX *ctx) { cipher_free_(ctx); }
```

Comment:
Agreed.  And it ticked me off.  The code already compensates.  Please see:

https://github.com/facebook/rocksdb/blob/09d5080dc80648bcc4e2520cc5fb72d842961e6e/util/library_loader.cc#L96

The code simply maps the ENV_CIPHER_CTX_destroy into the cipher_free_ pointer since parameters are same.

Reply:
OpenSSL 1.0.2 doesn't seems to provide `EVP_CIPHER_CTX_destroy` as well. With 1.0.2 the ctx can only be created on stack. https://www.openssl.org/docs/man1.0.2/man3/EVP_EncryptInit_ex.html

Reply:
Ok.  That would actually simplify the current code.  Using thread_local to get around recoding a bunch of stuff.  Keeps one context per thread ... which equates to one context per read or write call.

Reply:
I have pulled the source code to 1.0.1u.  It has functions _new, _free, and _cleanup.  I have pulled the source code to 1.0.2.u.  It has the functions _new, _free, and _cleanup.  Going to strongly suggest a documentation problem on their website.  My code is using _new, _free, and _reset.  ENV_CIPHER_CTX_reset() is a name change for _cleanup.  Source code uses the two equivalently inside _free and _init.  I believe the real action item here is to create code that will pull _cleanup as an alternative entry point to _reset.  

I was happily looking at maybe using a stack based copy of ENV_CIPHER_CTX.  There is no destructor and there are internal allocations that need clean up.  _new & _free might be safer in the long run, though calling _cleanup/_reset when done with stack struct accomplishes the same.  Looking at other alternatives too based upon review of ctrl128.c reference you posted elsewhere in this PR.